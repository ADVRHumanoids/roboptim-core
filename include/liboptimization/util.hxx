// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Useful procedures (templated) implementation.
 */
#ifndef OPTIMIZATION_UTIL_HXX
# define OPTIMIZATION_UTIL_HXX
# include <boost/variant/apply_visitor.hpp>

# include <liboptimization/function.hh>

namespace optimization
{
  namespace detail
  {
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x)
    {
      for (unsigned i = 0; i < jac.size1 (); ++i)
        {
          DerivableFunction::jacobian_t grad = c[i]->jacobian (x);
          for (unsigned j = 0; j < jac.size2 (); ++j)
            jac (i, j) = grad(0, j);
        }
    }


    struct GenVariantPrintVisitor : public boost::static_visitor<void>
    {
      GenVariantPrintVisitor (std::ostream& o)
        : o_ (o)
      {}

      template <typename T>
      void operator () (const T& value)
      {
        o_ << value;
      }

      template <typename T>
      void operator () (const T* value)
      {
        if (!value)
          o_ << "Null pointer";
        else
          o_ << *value;
      }

    private:
      std::ostream& o_;
    };

  }; // end of namespace detail.

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const boost::variant<T1, T2>& var)
  {
    detail::GenVariantPrintVisitor pv (o);
    boost::apply_visitor (pv, var);
    return o;
  }

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const std::vector<T>& vect)
  {
    typedef typename std::vector<T>::const_iterator citer_t;

    if (vect.empty ())
        return o << "Empty vector";

    citer_t it = vect.begin ();
    o << *it;
    ++it;

    for (; it != vect.end (); ++it)
      o << ", " << *it;
    return o;
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::pair<T1, T2>& p)
  {
    return o << "(" << p.first << ", " << p.second << ")";
  }

}; // end of namespace optimization.

#endif //! OPTIMIZATION_UTIL_HXX
