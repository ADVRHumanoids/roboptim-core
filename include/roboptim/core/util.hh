// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Useful procedures declaration.
 */
#ifndef ROBOPTIM_CORE_UTIL_HH
# define ROBOPTIM_CORE_UTIL_HH
# include <boost/variant.hpp>
# include <roboptim/core/derivable-function.hh>

namespace roboptim
{
  namespace detail
  {
    /// \internal
    /// Copy the content of a uBLAS vector into a C array.
    void vector_to_array (Function::value_type* dst, const Function::vector_t& src);

    /// \internal
    /// Copy the content of a C array into a uBLAS vector.
    void array_to_vector (Function::vector_t& dst, const Function::value_type* src);

    /// \internal
    /// Merge gradients from several functions (each gradient is a line).
    /// The first line of the jacobian is the only one used.
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x);
  } // end of namespace detail.

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream&, const boost::variant<T1, T2>&);

  template <typename T>
  std::ostream& operator<< (std::ostream&, const std::vector<T>&);

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream&, const std::pair<T1, T2>&);
} // end of namespace roboptim.

# include <roboptim/core/util.hxx>
#endif //! ROBOPTIM_CORE_UTIL_HH
