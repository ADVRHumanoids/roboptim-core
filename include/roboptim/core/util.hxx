// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_UTIL_HXX
# define ROBOPTIM_CORE_UTIL_HXX
# include <roboptim/core/function.hh>

namespace roboptim
{
  namespace detail
  {
    template <typename T>
    void
    jacobian_from_gradients (DerivableFunction::matrix_t& jac,
                             const std::vector<const T*>& c,
                             const DerivableFunction::vector_t& x)
    {
      for (unsigned i = 0; i < jac.size1 (); ++i)
        {
          DerivableFunction::jacobian_t grad = c[i]->jacobian (x);
          for (unsigned j = 0; j < jac.size2 (); ++j)
            jac (i, j) = grad(0, j);
        }
    }
  } // end of namespace detail.

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const std::vector<T>& vect)
  {
    typedef typename std::vector<T>::const_iterator citer_t;

    if (vect.empty ())
        return o << "Empty vector";

    citer_t it = vect.begin ();
    o << *it;
    ++it;

    for (; it != vect.end (); ++it)
      o << ", " << *it;
    return o;
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::pair<T1, T2>& p)
  {
    return o << "(" << p.first << ", " << p.second << ")";
  }

}; // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_UTIL_HXX
