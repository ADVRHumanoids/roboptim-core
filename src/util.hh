// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \file src/util.hh
 *
 * \brief Useful procedures declaration.
 */
#ifndef OPTIMIZATION_UTIL_HH
# define OPTIMIZATION_UTIL_HH
# include <function.hh>
# include <problem.hh>

namespace optimization
{
  namespace detail
  {
    /// Copy the content of a uBLAS vector into a C array.
    void vector_to_array (Function::value_type* dst, const Function::vector_t& src);

    /// Copy the content of a C array into a uBLAS vector.
    void array_to_vector (Function::vector_t& dst, const Function::value_type* src);

    /// Merge gradients from several functions (each gradient is a line).
    void jacobian_from_gradients (Function::matrix_t&,
                                  const Problem::constraints_t&,
                                  const Function::vector_t&);

  }; // end of namespace detail.
}; // end of namespace optimization.

#endif //! OPTIMIZATION_UTIL_HH
