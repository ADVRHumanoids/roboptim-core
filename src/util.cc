// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "debug.hh"

#include <cstring>
#include <boost/numeric/ublas/matrix.hpp>

#include "roboptim/core/util.hh"

namespace roboptim
{
  namespace detail
  {
    namespace ublas = boost::numeric::ublas;

    void
    vector_to_array (Function::value_type* dst, const Function::vector_t& src)
    {
      if (src.empty ())
	return;
      memcpy (dst, &src[0], src.size () * sizeof (Function::value_type));

      // NaN != NaN, handle this case.
      for (std::size_t i = 0; i < src.size (); ++i)
	if (src[i] != src[i])
	  assert (dst[i] != dst[i]);
	else
	  assert (dst[i] == src[i]);
    }

    void
    array_to_vector (Function::vector_t& dst, const Function::value_type* src)
    {
      if (dst.empty ())
	return;
      memcpy (&dst[0], src, dst.size () * sizeof (Function::value_type));

      // NaN != NaN, handle this case.
      for (std::size_t i = 0; i < dst.size (); ++i)
	if (src[i] != src[i])
	  assert (dst[i] != dst[i]);
	else
	  assert (dst[i] == src[i]);
    }
  } // end of namespace detail.
} // end of namespace roboptim.
