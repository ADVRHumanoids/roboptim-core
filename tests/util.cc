// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>

#include "common.hh"

#include <roboptim/core/io.hh>
#include <roboptim/core/util.hh>

using namespace roboptim;

int run_test ()
{
  // Check vector display.
  std::vector<int> v;

  std::cout << v << std::endl;

  v.push_back (1);
  std::cout << v << std::endl;

  return 0;
}

GENERATE_TEST ()
