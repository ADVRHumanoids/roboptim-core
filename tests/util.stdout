empty vector
1
(42, foo)
{}
{bar: 1337, foo: 42}
int
class roboptim::GenericFunction<struct roboptim::EigenMatrixDense>
[5,5]((0,0,0,0,0), (0,1,2,3,4), (0,2,4,6,8), (0,3,6,9,12), (0,4,8,12,16))
[5,5]((0,0,0,0,0), (0,1,2,3,4), (0,2,4,6,8), (0,3,6,9,12), (0,4,8,12,16))
[4](1,2,3,4)
[4](1,2,3,4)
[5,5]((0,3,0,0,0), (0,0,0,0,0), (0,12,0,42,0), (0,0,0,1,0), (0,0,0,0,0))
[5,5]((0,3,0,0,0), (0,0,0,0,0), (0,4,0,5,0), (0,0,0,6,0), (0,0,0,0,0))
[3,3]((12,0,42), (0,0,1), (0,0,0))
[5,5]((0,3,0,0,0), (0,0,0,0,0), (0,12,0,42,0), (0,0,0,1,0), (0,0,0,0,0))
[5,5]((0,6,0,0,0), (0,0,0,0,0), (0,24,0,84,0), (0,0,0,2,0), (0,0,0,0,0))
[5,5]((0,12,0,0,0), (0,0,0,0,0), (0,48,0,168,0), (0,0,0,4,0), (0,0,0,0,0))
[10](1,0.1,0.01,0.001,0.0001,1e-005,1e-006,1e-007,1e-008,1e-009)
[10](1,0.1,0.01,0.001,0.0001,1e-005,1e-006,1e-007,1e-008,0)
[4,4]((1,0.1,0.01,0.001), (0.0001,1e-005,1e-006,1e-007), (1e-008,1e-009,1e-010,1e-011), (1e-012,1e-013,1e-014,1e-015))
[4,4]((1,0.1,0.01,0.001), (0.0001,1e-005,1e-006,1e-007), (1e-008,0,0,0), (0,0,0,0))
"aaa" split with 'a': , , 
"ababa" split with 'b': a, a, a
"this is a test" split with ' ': this, is, a, test
"" split with '\n': empty vector
"an awesome
multiline test
with 3 lines!" split with '\n': an awesome, multiline test, with 3 lines!
Not expecting a FPE... done!
